File: GILua/src/main/java/org/anime_game_servers/gi_lua/models/Position.java
package org.anime_game_servers.gi_lua.models;

import com.github.davidmoten.rtreemulti.geometry.Point;
import org.anime_game_servers.core.gi.models.Vector;

public interface Position extends Vector {
    float getX();
    float getY();
    float getZ();


    default Point toPoint() {
        return Point.create(getX(),getY(),getZ());
    }

    /**
     * To XYZ array for Spatial Index
     */
    default double[] toDoubleArray() {
        return new double[]{ getX(),getY(),getZ()};
    }
    /**
     * To XZ array for Spatial Index (Blocks)
     */
    default double[] toXZDoubleArray() {
        return new double[]{getX(),getZ()};
    }
}

File: GILua/src/main/java/org/anime_game_servers/gi_lua/models/Position.kt
package org.anime_game_servers.gi_lua.models

import com.github.davidmoten.rtreemulti.geometry.Point
import org.anime_game_servers.core.gi.models.Vector

interface Position : Vector {
    override fun getX(): Float
    override fun getY(): Float
    override fun getZ(): Float


    fun toPoint(): Point {
        return Point.create(getX().toDouble(), getY().toDouble(), getZ().toDouble())
    }

    /**
     * To XYZ array for Spatial Index
     */
    override fun toDoubleArray(): DoubleArray {
        return doubleArrayOf(getX().toDouble(), getY().toDouble(), getZ().toDouble())
    }

    /**
     * To XZ array for Spatial Index (Blocks)
     */
    override fun toXZDoubleArray(): DoubleArray {
        return doubleArrayOf(getX().toDouble(), getZ().toDouble())
    }
}
